### PR TITLE
[1LP][RFR] Add TaggableCollection, update containers smart tag tests

### PR DESCRIPTION
**Pull request by mshriver**
Rebasing a contribution from @shalomnaim1 from #7391 to update to our current master.

{{ pytest: --long-running --use-provider cmqe --use-provider onprem -vvvv cfme/tests/cloud_infra_common/test_tag_objects.py  cfme/tests/cloud_infra_common/test_tag_visibility.py }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 62090998ac17d0d6348aefdca52ebd16c012e25e:

`cfme/containers/provider/__init__.py`:
- :bangbang: Line 23:2: **E999** *SyntaxError: invalid syntax*
- :bangbang: Line 23:3: **E227** *missing whitespace around bitwise or shift operator*
- :bangbang: Line 23:7: **E225** *missing whitespace around operator*
- :bangbang: Line 26:3: **E225** *missing whitespace around operator*
- :bangbang: Line 26:5: **E225** *missing whitespace around operator*
- :bangbang: Line 26:7: **E225** *missing whitespace around operator*
- :bangbang: Line 27:1: **E402** *module level import not at top of file*
- :bangbang: Line 28:7: **E225** *missing whitespace around operator*

`cfme/containers/provider/openshift.py`:
- :bangbang: Line 8:2: **E999** *SyntaxError: invalid syntax*
- :bangbang: Line 8:3: **E227** *missing whitespace around bitwise or shift operator*
- :bangbang: Line 8:7: **E225** *missing whitespace around operator*
- :bangbang: Line 12:3: **E225** *missing whitespace around operator*
- :bangbang: Line 12:5: **E225** *missing whitespace around operator*
- :bangbang: Line 12:7: **E225** *missing whitespace around operator*
- :bangbang: Line 13:1: **E402** *module level import not at top of file*
- :bangbang: Line 14:7: **E225** *missing whitespace around operator*
- :bangbang: Line 91:3: **E227** *missing whitespace around bitwise or shift operator*
- :bangbang: Line 91:7: **E225** *missing whitespace around operator*
- :bangbang: Line 94:1: **E305** *expected 2 blank lines after class or function definition, found 0*
- :bangbang: Line 94:1: **E112** *expected an indented block*
- :bangbang: Line 94:3: **E225** *missing whitespace around operator*
- :bangbang: Line 94:5: **E225** *missing whitespace around operator*
- :bangbang: Line 94:7: **E225** *missing whitespace around operator*
- :bangbang: Line 95:1: **E302** *expected 2 blank lines, found 0*
- :bangbang: Line 98:1: **E305** *expected 2 blank lines after class or function definition, found 1*
- :bangbang: Line 98:1: **E112** *expected an indented block*
- :bangbang: Line 98:7: **E225** *missing whitespace around operator*

`cfme/tests/containers/test_smart_management.py`:
- :red_circle: Line 14:1: **F401** *'cfme.utils.log.logger' imported but unused*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by john-dupuy** [comment]
Can you please lint your PR? :grin: :grin: :grin:

**Comment by mshriver** [comment]
hey! its still in wiptest!

File: cfme/tests/containers/test_smart_management.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    pytest.mark.polarion('CMP-9870')(ContainersTestItem(
+        Project, 'CMP-9870', collection_obj="container_projects", keyword_name="name")),
+    pytest.mark.polarion('CMP-9854')(ContainersTestItem(
+        ContainersProvider, 'CMP-9854', collection_obj="containers_providers",
+        keyword_name="name"))
+]
+
+
+def get_entity(collection):
+
```
Comment (digitronik) [inline cfme/tests/containers/test_smart_management.py]: I hate blank lines at start without doc string. but totally optional :)

File: cfme/tests/containers/test_smart_management.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    # Map all object by name
+    mapping = {item.name: item for item in collection.all()}
+
+    # Filter only the unique items
+    unique_items = [key for key in set(mapping.keys()) if mapping.keys().count(key) == 1]
```
Comment (digitronik) [inline cfme/tests/containers/test_smart_management.py]: Q: why you need to convert `dict_keys` to `set`?

File: cfme/tests/containers/test_smart_management.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    # If there is at least one  unique item
```
Comment (digitronik) [inline cfme/tests/containers/test_smart_management.py]: Two time space between one and unique

File: cfme/tests/containers/test_smart_management.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    obj_inst = (get_clean_entity(entity=provider)
+                if getattr(test_item.obj, 'one_of', BaseProvider.one_of)(ContainersProvider)
+                else get_clean_entity(collection=obj_collection))
+
+    # Config random tag for object
+    tag = obj_collection.add_tag([obj_inst], keyword_name=test_item.keyword_name)
+
+    all_tags = wait_for(obj_inst.get_tags, num_sec=30, delay=5).out
+
+    # Validate tag wsa set successfully
```
Comment (digitronik) [inline cfme/tests/containers/test_smart_management.py]: Required:
`typo`

File: cfme/tests/containers/test_smart_management.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    entity.remove_tags(tags)
+    return entity
+
+
+@pytest.mark.parametrize('test_item', TEST_ITEMS,
+                         ids=[ContainersTestItem.get_pretty_id(ti) for ti in TEST_ITEMS])
+@pytest.mark.meta(blockers=[BZ(1609541)])
+def test_smart_management_add_tag(provider, appliance, test_item):
+
+    # validate no tag set to project
```
Comment (digitronik) [inline cfme/tests/containers/test_smart_management.py]: Optional:
Better to call `entity` or `object`. Project scope is much as per me.

File: cfme/tests/containers/test_smart_management.py (hunk 1 above)
Comment (mshriver) [reply cfme/tests/containers/test_smart_management.py]: A lot of this is not my original commits, btw. will clean this stuff up though, thanks for pointing out.

File: cfme/tests/containers/test_smart_management.py (hunk 2 above)
Comment (mshriver) [reply cfme/tests/containers/test_smart_management.py]: Didn't write or modify this line in pulling it from older commits. not sure about why this is done this way.

Comment (mshriver) [reply cfme/tests/containers/test_smart_management.py]: Okay, so reading this a bit more, the logic is kinda screwy here and needs work. Thanks for pointing this out @digitronik 

I'm going to rewrite this method, or at least these first two statements

File: cfme/tests/containers/test_smart_management.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    entity.remove_tags(tags)
+    return entity
+
+
+@pytest.mark.parametrize('test_item', TEST_ITEMS,
+                         ids=[ContainersTestItem.get_pretty_id(ti) for ti in TEST_ITEMS])
+@pytest.mark.meta(blockers=[BZ(1609541)])
+def test_smart_management_add_tag(provider, appliance, test_item):
+
+    # validate no tag set to object
```
Comment (ganeshhubale) [inline cfme/tests/containers/test_smart_management.py]: @mshriver 
Can you please tell me, why don't we have ```polarion``` tags for this test case?

File: cfme/tests/containers/test_smart_management.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    obj_collection = getattr(appliance.collections, test_item.collection_obj)
+    obj_inst = (get_clean_entity(entity=provider)
+                if getattr(test_item.obj, 'one_of', BaseProvider.one_of)(ContainersProvider)
```
Comment (ganeshhubale) [inline cfme/tests/containers/test_smart_management.py]: @mshriver 
- I think we only need ```appliance``` [here](https://github.com/ManageIQ/integration_tests/pull/8544/files#diff-8a9c95d49623f7876fc412cc1eed90f7R89). 
- Will it be fine; if we go with ```provider.appliance.collections```?

**Comment by jawatts** [comment]
CMQE no longer has a smart management test. I think this should go under `test_tags`. Moving to WIPTEST

File: cfme/tests/containers/test_smart_management.py (hunk 6 above)
Comment (mshriver) [reply cfme/tests/containers/test_smart_management.py]: @ganeshhubale good points, thanks!

This file is actually getting removed after some discussion with @jawatts. These tests are covered in other modules, and this ends up being a duplicate. 

It got added "back" basically, when I pulled these old commits back from the dead.

**Comment by jawatts** [comment]
Failures in PRT not due to changes

**Pull request merged**